### DISCUSSION TITLE
[@types/markdown-it] `lib/*` exports `.js` files but library has `.mjs` file extensions

### DISCUSSION THREAD

**Discussion by runarberg** [DefinitelyTyped/DefinitelyTyped#68802] [Issues with a @types package]
When I import files from `markdown-it/lib` in JavaScript I need to specify the `.mjs` extension. However the Definitely Typed declarations are all of the `.d.ts` kind, as opposed to the `.d.mts`. Meaning that every time I do something like:

```js
import Token from "markdown-it/lib/token.mjs";
```

It correctly imports the class in JavaScript, however when I run `tsc --noEmit` I get this error:

```
my-module.js:1:19 - error TS7016: Could not find a declaration file for module 'markdown-it/lib/token.mjs'. '/path/to/my-module/node_modules/markdown-it/lib/token.mjs' implicitly has an 'any' type.
  If the 'markdown-it' package actually exposes this module, consider sending a pull request to amend 'https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/markdown-it'
```

If I change the import statement to use `.js` extension instead:

```js
import Token from "markdown-it/lib/token.js";
```

Now the type check passes but the script fails on unresolved import.

```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/path/to/module/node_modules/markdown-it/lib/token.js' imported from /path/to/module/my-module.js
```

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment]
Hey, it's because DT version of the types was never migrated to MJS module:
https://github.com/markdown-it/markdown-it/commit/4949a10120d101bdca48bf1b13f9e790d6f2049e#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR12
(introduced in version 14)

**Reply by runarberg** [reply]
So in short, `@types/markdown-it@14` was hasn’t been released. I’m not super familiar with `.d.ts` files but would it be enough then to rename all the files from `.d.ts` to `.d.mts` and then the issue is resolved?

Is there a more elegant/idiomatic way?

**Reply by peterblazejewicz** [reply]
the types of DT package in package.json, should reflect being module as well.

**Reply by runarberg** [reply]
Should I add: `"type": "module"` to the DT `package.json` even when the reflected `package.json` at Markdown-it does not have that field?

**Reply by peterblazejewicz** [reply]
yes, you're correct, I was not formally correct (the problem stays). This needs to be reflected in types definition:
https://github.com/markdown-it/markdown-it/blob/master/package.json#L15-L25
(with changed exports/imports)

**Reply by runarberg** [reply]
I tried renaming all the files (along with changing `require()` to import statements as well as `export =` to export statements) and matching the export fields in the `package.json`. However I could not get the tests to pass, no matter how I tried.

I took a look at how other packages do this and it seems like the keep the commonjs files and simply add an `.d.mts` which exports from it. I’m not sure this is actually how markdown-it exports their modules (they only export the `MarkdownIt` constructor from `/dist/index.cjs.js` as a commonjs module, but all their other files as es6 modules).

**Reply by runarberg** [reply]
Actually, that doesn’t seem to work either, then I get a lint error:

> Module "[...]" can only be default-imported using the 'esModuleInterop' flag

but the `esModuleInterop` flag is [explicitly forbidden](https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#esmoduleinteropallowsyntheticdefaultimports) by the DT README.

**Reply by runarberg** [reply]
Here is my attempt so far https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...runarberg:DefinitelyTyped:master

But I still can’t get the tests to work.
